File: backend/models/scheduler.py
"""
计划任务相关数据模型
"""
from tortoise import fields
from tortoise.models import Model


class SchedulerTask(Model):
    """计划任务表"""
    id = fields.IntField(pk=True)
    task_key = fields.CharField(max_length=50, unique=True, description="任务KEY")
    task_name = fields.CharField(max_length=100, description="任务名称")
    task_type = fields.CharField(max_length=50, default="monitor", description="任务类型")
    trigger_type = fields.CharField(max_length=20, default="cron", description="触发器类型: cron/interval/date")
    trigger_config = fields.CharField(max_length=200, description="触发器配置(cron表达式等)")
    job_path = fields.CharField(max_length=200, description="任务执行路径(模块.函数)")
    job_args = fields.TextField(null=True, description="任务参数JSON")
    job_kwargs = fields.TextField(null=True, description="任务关键字参数JSON")
    is_enabled = fields.BooleanField(default=True, description="是否启用")
    is_running = fields.BooleanField(default=False, description="是否正在运行")
    description = fields.TextField(null=True, description="任务描述")
    last_run_time = fields.DatetimeField(null=True, description="上次执行时间")
    next_run_time = fields.DatetimeField(null=True, description="下次执行时间")
    last_run_status = fields.CharField(max_length=20, null=True, description="上次执行状态")
    last_run_message = fields.TextField(null=True, description="上次执行消息")
    run_count = fields.IntField(default=0, description="执行次数")
    created_at = fields.DatetimeField(auto_now_add=True)
    updated_at = fields.DatetimeField(auto_now=True)

    class Meta:
        table = "scheduler_tasks"
        indexes = [
            ("task_key",),
            ("is_enabled",),
            ("task_type",),
        ]

    def __str__(self):
        return f"{self.task_key} - {self.task_name}"


class TaskLog(Model):
    """任务执行日志表"""
    id = fields.IntField(pk=True)
    task_id = fields.IntField(description="任务ID")
    task_key = fields.CharField(max_length=50, description="任务KEY")
    task_name = fields.CharField(max_length=100, description="任务名称")
    job_group = fields.CharField(max_length=50, null=True, description="任务组")
    job_executor = fields.CharField(max_length=50, null=True, description="执行器")
    invoke_target = fields.CharField(max_length=200, description="执行目标")
    job_args = fields.TextField(null=True, description="位置参数")
    job_kwargs = fields.TextField(null=True, description="关键字参数")
    job_trigger = fields.CharField(max_length=100, null=True, description="触发器")
    job_message = fields.TextField(null=True, description="执行消息")
    status = fields.BooleanField(description="执行状态")
    exception_info = fields.TextField(null=True, description="异常信息")
    start_time = fields.DatetimeField(description="开始时间")
    end_time = fields.DatetimeField(null=True, description="结束时间")
    duration = fields.IntField(null=True, description="执行耗时(毫秒)")
    created_at = fields.DatetimeField(auto_now_add=True)

    class Meta:
        table = "task_logs"
        indexes = [
            ("task_id",),
            ("task_key",),
            ("status",),
            ("created_at",),
        ]

    def __str__(self):
        return f"{self.task_key} - {self.status}"

File: frontend/src/api/scheduler.ts
/**
 * 计划任务相关API
 */
import request from '@/utils/request'

// ==================== 类型定义 ====================

export interface SchedulerTask {
  id: number
  task_key: string
  task_name: string
  task_type: string
  trigger_type: string
  trigger_config: string
  job_path: string
  is_enabled: boolean
  is_running: boolean
  description?: string
  last_run_time?: string
  next_run_time?: string
  last_run_status?: string
  run_count: number
}

export interface TaskLog {
  id: number
  task_key: string
  task_name: string
  status: boolean
  job_message?: string
  exception_info?: string
  start_time?: string
  end_time?: string
  duration?: number
  created_at?: string
}

export interface CronExample {
  expr: string
  desc: string
}

export interface CronPreviewResult {
  cron_expr: string
  valid: boolean
  next_runs?: string[]
  error?: string
}

// ==================== 任务管理接口 ====================

/** 获取计划任务列表 */
export async function fetchSchedulerTasks(params?: {
  task_type?: string
  is_enabled?: boolean
}): Promise<SchedulerTask[]> {
  return request.get('/api/scheduler/tasks', { params })
}

/** 创建计划任务 */
export async function createSchedulerTask(data: {
  task_key: string
  task_name: string
  task_type?: string
  trigger_type?: string
  trigger_config: string
  job_path: string
  job_args?: string
  job_kwargs?: string
  description?: string
}): Promise<{ id: number }> {
  return request.post('/api/scheduler/tasks', data)
}

/** 更新计划任务 */
export async function updateSchedulerTask(id: number, data: {
  task_name?: string
  trigger_type?: string
  trigger_config?: string
  is_enabled?: boolean
  description?: string
}): Promise<void> {
  return request.put(`/api/scheduler/tasks/${id}`, data)
}

/** 删除计划任务 */
export async function deleteSchedulerTask(id: number): Promise<void> {
  return request.delete(`/api/scheduler/tasks/${id}`)
}

// ==================== 任务控制接口 ====================

/** 启动任务 */
export async function startTask(id: number): Promise<void> {
  return request.post(`/api/scheduler/tasks/${id}/start`)
}

/** 停止任务 */
export async function stopTask(id: number): Promise<void> {
  return request.post(`/api/scheduler/tasks/${id}/stop`)
}

/** 立即执行任务 */
export async function runTaskNow(id: number): Promise<void> {
  return request.post(`/api/scheduler/tasks/${id}/run`)
}

// ==================== 任务日志接口 ====================

/** 获取任务日志 */
export async function fetchTaskLogs(taskId: number, limit?: number): Promise<TaskLog[]> {
  return request.get(`/api/scheduler/tasks/${taskId}/logs`, { params: { limit } })
}

/** 获取所有日志 */
export async function fetchAllLogs(params?: {
  task_key?: string
  status?: boolean
  limit?: number
}): Promise<TaskLog[]> {
  return request.get('/api/scheduler/logs', { params })
}

// ==================== Cron辅助接口 ====================

/** 预览Cron表达式 */
export async function previewCron(cronExpr: string): Promise<CronPreviewResult> {
  return request.post('/api/scheduler/cron/preview', { cron_expr: cronExpr })
}

/** 获取Cron示例 */
export async function fetchCronExamples(): Promise<CronExample[]> {
  return request.get('/api/scheduler/cron/examples')
}

// ==================== 调度器状态接口 ====================

/** 获取调度器状态 */
export async function fetchSchedulerStatus(): Promise<{
  status: string
  job_count: number
  jobs: any[]
}> {
  return request.get('/api/scheduler/status')
}

/** 初始化预设任务 */
export async function initDefaultTasks(): Promise<void> {
  return request.post('/api/scheduler/init')
}